crates/pathfinder/src/cairo/ext_py/sub_process.rs

```diff
@@ -217,10 +225,79 @@ async fn spawn(
                     }
                 }
 
-                trace!("{:?}", buffer.trim());
+                let buffer = buffer.trim();
+
+                struct InvalidLevel;
+
+                // first one is the level selector, assuming this is code controlled by us
+                let mut chars = buffer.chars();
+                let level = match chars.next() {
+                    Some('0') => Ok(tracing::Level::ERROR),
+                    Some('1') => Ok(tracing::Level::WARN),
+                    Some('2') => Ok(tracing::Level::INFO),
+                    Some('3') => Ok(tracing::Level::DEBUG),
+                    Some('4') => Ok(tracing::Level::TRACE),
+                    Some(_) => Err(InvalidLevel),
+                    None => {
+                        // whitespace only line
+                        continue;
+                    }
+                };
+
+                let (level, displayed) = if let Ok(level) = level {
+                    let rem = chars.as_str();
+                    // if we treat the thing as json, we can easily get multiline messages, which most
+                    // are
+                    let displayed = if rem.starts_with('"') {
+                        serde_json::from_str::<String>(rem)
+                            .ok()
+                            .map(std::borrow::Cow::Owned)
```

Comment:
Feels quite extra to invoke serde_json here, but... (looks at the spec) ... Tempting to do this in place. I'll leave that to another PR :)